onnxruntime/core/graph/graph.cc

```diff
@@ -2762,10 +2762,9 @@ const ONNX_NAMESPACE::TensorProto* Graph::GetConstantInitializer(const std::stri
 
 #if !defined(ORT_MINIMAL_BUILD)
 void Graph::AddValueInfo(const NodeArg* new_value_info) {
-  for (const auto* info : value_info_) {
-    ORT_ENFORCE(info->Name() != new_value_info->Name(), "Error: trying to add an existing value info.");
-  }
-  value_info_.push_back(new_value_info);
+  NodeArg* node_arg = GetNodeArg(new_value_info->Name());
```

Comment:
Can you set a custom comparator for the unordered_set so that you don't have to do the existence set manually? Then you can simply check the return value of the insert method.

Reply:
The check is to check if the NodeArg is in the graph, which is on another unordered_map, not in the value_info_, will adding custom comparator to the value_info_ set help this?

Reply:
could it be a comparator struct lives under Graph, so that it have access to GetNodeArg()?

Reply:
Not sure why a call to GetNodeArg() is needed. You already have new_value_info which is a NodeArg*. The previous lines 2765-2767 didn't need GetNodeArg to check for uniqueness.

Reply:
Because the caller may create a NodeArg outside of the graph and try to add it here. It's doable but I am thinking this is logically wrong. The previous implementation didn't check this. I can remove the GetNodeArg(), then yes I will need a comparator to check the uniqueness of the name. But if at the beginning we require the NodeArg must from inside the graph, then we can use the pointer as key directly. Or maybe the best signature of this function is not a pointer of new NodeArg, it's better a string which is the name of the NodeArg?

Reply:
Ok, I see. But the NodeArg's coming from the graph node_args_ are already unique since node_args_ is a map keyed by name. So, if you're calling GetNodeArg first to ensure only values from the graph are added, is there still a need to make value_info_ an unordered_set?

Reply:
When caller adds a pointer of NodeArg, we don't know if it's already in the value_info_ container, right? It's unique in the node_arg_ map. If value_info_ is vector, I still need to O(n) to check if the pointer is inside it by comparing the pointer. Make it to set change it to O(1).